# Linear Regression

### Data Prep

In [3]:
import pandas as pd
data_frame = pd.read_csv("acs2017_census_tract_data.csv")
copy = data_frame.copy()
copy.drop(labels="Poverty", axis="columns", inplace=True)
copy.drop(labels="State", axis="columns", inplace=True)
copy.drop(labels="County", axis="columns", inplace=True)
copy.drop(labels="Men", axis="columns", inplace=True)
copy.drop(labels="Women", axis="columns", inplace=True)
copy.drop(labels="Hispanic", axis="columns", inplace=True)
copy.drop(labels="White", axis="columns", inplace=True)
copy.drop(labels="Native", axis="columns", inplace=True)
copy.drop(labels="Pacific", axis="columns", inplace=True)
copy.drop(labels="Asian", axis="columns", inplace=True)
copy.drop(labels="VotingAgeCitizen", axis="columns", inplace=True)
copy.drop(labels="ChildPoverty", axis="columns", inplace=True)
copy.drop(labels="Black", axis="columns", inplace=True)
copy.dropna(axis="index", how="any", inplace=True)
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(copy, test_size=0.2, random_state=123)


### Initial set of features
I would like my x features to be Professional and Office. They are both percentage values as in percentage of the population of a county works in these fields. I would like my y feature to be income. This is the median income of the county and is our predicted value. 

In [12]:
working_set = train_set.copy()
x_data = working_set[["Professional", "Office"]]
y_data = working_set["Income"]

from sklearn.linear_model import LinearRegression
reg = LinearRegression()
reg.fit(x_data, y_data)
print("R2: ", reg.score(x_data, y_data))

from sklearn.metrics import mean_squared_error
y_pred = reg.predict(x_data)
mse = mean_squared_error(y_data, y_pred)
print("MSE: ",  mse)

R2:  0.5438220655393842
MSE:  425940188.33046246


#### Comments:
This data seemed to do alright, 54.38% of the variation in y can be explained by the variation in x. Our MSE is high, perhaps we could get it lower by adding more predictors.

In [10]:
x_data2 = working_set[["Professional", "Office", "Service", "Construction"]]

reg2 = LinearRegression()
reg2.fit(x_data2, y_data)
print("R2: ", reg2.score(x_data2, y_data))

y_pred2 = reg.predict(x_data2)
mse2 = mean_squared_error(y_data, y_pred2)
print("MSE: ",  mse2)

R2:  0.5618306831720081
MSE:  409125271.59171057


### Comments:
The R2 value has increased and the MSE has decreased. This is good for the model although it wasn't much of an improvement. 

In [13]:
working_set2 = test_set.copy()
test_x = working_set2[["Professional", "Office", "Service", "Construction"]]
test_y = working_set2["Income"]

print("R2: ", reg2.score(test_x, test_y))
y_pred3 = reg2.predict(test_x)
mse3 = mean_squared_error(test_y, y_pred3)
print("MSE: ", mse3)

R2:  0.5709985079497797
MSE:  394633961.790895


### Comments: 
The test set performed well, we were able to predict values more often in the test set than in the training set and the MSE was lower also. 